### PR TITLE
RESTEasy Reactive

### PR DESCRIPTION
**Pull request by FroMage**
Initial implementation of the new Quarkus REST extension.

This is a fresh from-the-grounds-up implementation of JAX-RS that is almost spec-compliant, focuses on build-time initialisation, vertx runtime, asynchronous, and extensions to the spec.

### CODE REVIEW

**Review by geoand — APPROVED** [review]

**Comment by geoand** [comment]
Let's not merge this just yet, but let's indeed get some CI feedback

**Comment by gsmet** [comment]
@geoand I have no idea how it's best to integrate this?

457 commits look like horrible if we have to somehow "manipulate" this feature later? But I don't know if it makes more sense to keep it that way or to squash at least part of it?

What have me worried is that I see some "fix formatting" commits, which means, for some of the commits, the build doesn't pass and that's something we're trying to avoid to allow bissecting.

**Comment by gsmet** [comment]
So just to be clear, given the history, this will break any tentative to bissect Quarkus from before this merge to after this merge.

**Comment by stuartwdouglas** [comment]
I think the bisect issue can be worked around with a little bit of bash/git magic. If we do end up with a situation where this is a problem I think you can do:

git checkout old-quarkus-rest-sha (parent of this merge commit)
git cherry old-quarkus-master-sha (other parent)

This will give you a list of all the commits that were added, and then I think you can use a bash for loop to run 'git bisect skip sha' to mark them as unstestable. Its slightly inconvenient, but bisect does not come up that often, and once you have the bash code anyone can re-use it if it happens again.

**Comment by stuartwdouglas** [comment]
`git cherry old-quarkus-master-sha old-quarkus-rest-sha | cut -f 2 -d\ | xargs git bisect skip`

Should do it I think

**Comment by geoand** [comment]
@FroMage the SSE test is flaky

**Comment by gsmet** [comment]
@stuartwdouglas there are several issues:

- the formatting issues should be fixable easily with something like http://kamalmarhubi.com/blog/2016/03/08/git-rebase-exec-make-sure-your-tests-pass-at-each-commit-and-other-rebase-goodies/ and that would avoid having the tree in a bad state and at least could remove these useless commits
- spreading a feature over 400+ commits is going to be a major pain. Just having a look at the commit list in the UI will be totally unusable for a while because of this. And I can't imagine all the pain we will have because of it. Just to give an idea, this PR is ~3% of the number of commits since the very beginning of Quarkus.
- I don't think the TCK should be merged in the main tree. It should be a separate project. Rebuilding the TCK for each CI run is a total waste of resource.
- Good luck to get a proper review given the GH UI is barely usable because of the number of files. Just externalizing the TCK could help with this.

I don't expect people to spend 2 weeks cleaning this commit tree but I think spending a day or two to make this more digestable would be worth it given we will have this in the tree forever.

I'm pretty sure my comments will be outruled but I certainly can't let this thing be merged as is without saying it out loud.

**Comment by gsmet** [comment]
Just to apologize if you took my comment as a criticism of the quality of the work and congrats on the achievement.

If it wasn't clear, this was not a comment about the quality of the work, I'm really not able to judge that and I certainly trust you three to do amazing work in this area. It was just about the form of the PR that makes it nearly impossible to review. I have no idea if it's possible to do better but I think we should at least spend a reasonable amount to time to try to make it reviewable and have a more digestible history - for the review and also for our future selves who will have to browse the Quarkus history.

File: extensions/quarkus-rest-jackson/runtime/src/main/resources/quarkus-extension.yaml
Comment (machi1990) [inline extensions/quarkus-rest-jackson/runtime/src/main/resources/quarkus-extension.yaml:2]:
```suggestion
name: "Quarkus REST Jackson"
```

**Comment by Postremus** [comment]
Some points from my preliminary testing:
* No endpoints are included in the /openapi file.
* The NotFoundExceptionMapper is not included with this extension -> no nice error page in dev mode when the url is wrong.

**Comment by geoand** [comment]
Thanks for checking.

Neither of those was the case last time I checked.

OpenAPI should work and the nice not found page should also work in dev mode.
But I'll check again soon

**Comment by geoand** [comment]
@Postremus I just tried this again and both `/openapi` and the custom error page for HTTP 404 in dev mode work.

I tried the openapi quickstart. How are you able to reproduce the problems you mentioned?

**Comment by geoand** [comment]
@gastaldi the script that generates the build items doc seems to failing... 

```
2020-10-29T06:09:25.7530820Z [INFO] --- jbang-maven-plugin:0.0.5:run (generate-build-item-doc) @ quarkus-documentation ---
2020-10-29T06:09:25.8389680Z [WARNING] JBang not found. Downloading version 0.47.1
2020-10-29T06:09:26.0884791Z [WARNING] No signatures were supplied, skipping file validation
2020-10-29T06:09:26.0895884Z [INFO] Got from cache: /home/runner/.m2/repository/.cache/download-maven-plugin/jbang-0.47.1.zip_79deed705ef19e0fc580b11662b53954
2020-10-29T06:09:26.7742027Z [INFO] Using JBang v.0.47.1
2020-10-29T06:09:26.7742705Z 
2020-10-29T06:09:27.5428663Z [INFO] [jbang] Resolving dependencies...
2020-10-29T06:09:27.8288134Z [INFO] [jbang]     Resolving info.picocli:picocli:4.5.0...Done
2020-10-29T06:09:27.8857609Z [INFO] [jbang]     Resolving org.jboss.forge.roaster:roaster-jdt:2.22.2.Final...Done
2020-10-29T06:09:27.9076085Z [INFO] [jbang]     Resolving org.eclipse.collections:eclipse-collections:10.4.0...Done
2020-10-29T06:09:27.9171753Z [INFO] [jbang]     Resolving org.yaml:snakeyaml:1.27...Done
2020-10-29T06:09:27.9488713Z [INFO] [jbang]     Resolving io.fabric8:maven-model-helper:16...Done
2020-10-29T06:09:27.9499125Z [INFO] [jbang] Dependencies resolved
2020-10-29T06:09:28.0427906Z [INFO] [jbang] Building jar...
2020-10-29T06:09:32.4661204Z [INFO] org.jboss.forge.roaster.ParserException: Source does not represent a [JavaClassSource], instead was [JavaInterfaceImpl] - Cannot convert.
2020-10-29T06:09:32.4663906Z [INFO] 	at org.jboss.forge.roaster.Roaster.parse(Roaster.java:267)
2020-10-29T06:09:32.4665277Z [INFO] 	at org.jboss.forge.roaster.Roaster.parse(Roaster.java:285)
2020-10-29T06:09:32.4666722Z [INFO] 	at org.jboss.forge.roaster.Roaster.parse(Roaster.java:210)
2020-10-29T06:09:32.4668157Z [INFO] 	at quarkusbuilditemdoc.process(quarkusbuilditemdoc.java:107)
2020-10-29T06:09:32.4669353Z [INFO] 	at quarkusbuilditemdoc$1.visitFile(quarkusbuilditemdoc.java:97)
2020-10-29T06:09:32.4670484Z [INFO] 	at quarkusbuilditemdoc$1.visitFile(quarkusbuilditemdoc.java:93)
2020-10-29T06:09:32.4682837Z [INFO] 	at java.base/java.nio.file.Files.walkFileTree(Files.java:2724)
2020-10-29T06:09:32.4684389Z [INFO] 	at java.base/java.nio.file.Files.walkFileTree(Files.java:2796)
2020-10-29T06:09:32.4685662Z [INFO] 	at quarkusbuilditemdoc.collect(quarkusbuilditemdoc.java:93)
2020-10-29T06:09:32.4686879Z [INFO] 	at quarkusbuilditemdoc.call(quarkusbuilditemdoc.java:61)
2020-10-29T06:09:32.4688042Z [INFO] 	at quarkusbuilditemdoc.call(quarkusbuilditemdoc.java:37)
2020-10-29T06:09:32.4689328Z [INFO] 	at picocli.CommandLine.executeUserObject(CommandLine.java:1933)
2020-10-29T06:09:32.4690564Z [INFO] 	at picocli.CommandLine.access$1100(CommandLine.java:145)
2020-10-29T06:09:32.4693297Z [INFO] 	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2332)
2020-10-29T06:09:32.4699367Z [INFO] 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2326)
2020-10-29T06:09:32.4700344Z [INFO] 	at picocli.CommandLine$RunLast.handle(CommandLine.java:2291)
2020-10-29T06:09:32.4701613Z [INFO] 	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2159)
2020-10-29T06:09:32.4702715Z [INFO] 	at picocli.CommandLine.execute(CommandLine.java:2058)
2020-10-29T06:09:32.4703625Z [INFO] 	at quarkusbuilditemdoc.main(quarkusbuilditemdoc.java:50)
```

Any idea what's going on?

**Comment by stuartwdouglas** [comment]
I had an interface that ended with the BuildItem suffix, I will change it.

**Comment by geoand** [comment]
Ah, I see. 

Interesting how the doc script now acts as a safeguard against that 😎

File: extensions/quarkus-rest-jackson/runtime/src/main/resources/quarkus-extension.yaml
Comment (geoand) [reply extensions/quarkus-rest-jackson/runtime/src/main/resources/quarkus-extension.yaml]: :+1: . I'll change it when I go through the commit history and try to clean it up

**Comment by gastaldi** [comment]
@geoand @stuartwdouglas awesome! Does it make sense to have BuildItems as interfaces? The generator might need to be updated to support that then.

**Comment by geoand** [comment]
Stuart changed it IIUC

**Comment by geoand** [comment]
I just force pushed to the PR after cleaning up the Git history some

**Comment by geoand** [comment]
Also for those interested, the testsuite is now at https://github.com/quarkusio/quarkus-rest-testsuite (as to not polutte the main repo with a ton of otherwise useless code)

**Comment by geoand** [comment]
PR updated to use the Smallrye versions of `@Blocking` and `@NonBlocking`

**Comment by geoand** [comment]
Force-pushed in order to fix a conflict

**Pull request merged**
